src/server/svr_chk_owner.c

```diff
@@ -390,66 +390,94 @@ authenticate_user(struct batch_request *preq, struct connection *pcred)
  *		this checks the most conditions common to most job batch requests.
  *		It also returns a pointer to the job if found and the tests pass.
  *		If the request is for a single subjob or a range of subjobs (of an
- *		Job Array),  the return job pointer is to the parent Array Job.
+ *		Job Array), alsoe return pointer to the parent Array Job.
  * @par
  *		Depending on what the "jobid" identifies, the following is returned
- *		in the integer pointed to by rc:
+ *		in the integer pointed to by jobtype:
  *	   	IS_ARRAY_NO (0)       - for a regular job
  *	   	IS_ARRAY_ArrayJob (1) - for an Array Job
  *	   	IS_ARRAY_Single (2)   - for a single subjob
  *	   	IS_ARRAY_Range (3)    - for a range of  subjobs
  *
  * @param[in]	jobid	-	Job Id.
  * @param[in,out]	preq	-	job batch request
- * @param[out]	rc	-	Depending on what the "jobid" identifies,
+ * @param[out]	jobtype	-	Depending on what the "jobid" identifies,
  * 						the following is returned
  *						(0)       - for a regular job
  *						(1) - for an Array Job
  *						(2)   - for a single subjob
  *						(3)    - for a range of  subjobs
+ * @param[out]	jobstate	-	the state of the job
+ * @param[out]	ppjob	-	pointer to the job
+ * @param[out]	parent	-	pointer to the parent job of subjob
+ * @param[out]	poffset	-	offset of subjob in tracking table
  *
- * @return	job *
- * @retval	a pointer to the job	: if found and the tests pass.
- * @retval	NULL	: failed
+ * @return	int
+ * @retval	TRUE	: if found and the tests pass.
+ * @retval	FALSE	: test failed
  */
 
 
-job *
-chk_job_request(char *jobid, struct batch_request *preq, int *rc)
+int
+chk_job_request(char *jobid, struct batch_request *preq, int *jobtype, int *jobstate, job **ppjob, job **parent, int *poffset)
 {
-	int	 t;
+	int t;
+	int offset;
 	int	 histerr = 0;
 	job	*pjob;
 	int deletehist = 0;
 	char	*p1;
 	char	*p2;
 
+	if (ppjob == NULL || jobtype == NULL || jobstate == NULL ) {
+		/* good to add a debug log here */
+		return FALSE;
+	}
+
 	if (preq->rq_extend && strstr(preq->rq_extend, DELETEHISTORY))
 		deletehist = 1;
-	t = is_job_array(jobid);
-	if ((t == IS_ARRAY_NO) || (t == IS_ARRAY_ArrayJob))
-		pjob = find_job(jobid);		/* regular or ArrayJob itself */
+
+	t = *jobtype = is_job_array(jobid);
+
+	if (t != IS_ARRAY_Range)
```

Comment:
*jobtype itself can be used. No need of extra variable 't'. Apply this comment wherever possible.

Reply:
using the local variable t will help the compiler optimize and speed up the future access

Reply:
I must beg to differ. Consider the following code...
```
int
is_job_array(int jobid)
{
    return (jobid % 2);
}

int
chk_job_request(int jobid, int *jobtype)
{
    *jobtype = is_job_array(jobid);
    if (*jobtype != 0)
        *jobtype += jobid % 20;
    else
        *jobtype *= -1;
    if (*jobtype > 0)
        *jobtype *= 4;
    else
        *jobtype *= -5;
    return *jobtype;
}

int
main(void)
{
    int jobid, jobtype, ret;

    for (jobid=0 ; jobid < 1000000000; jobid++)
        ret = chk_job_request(jobid, &jobtype);
    return (jobtype % 5) + ret;
}
```

When compiled with -O2, I see the following...
```
$ time ./perf

real    0m1.075s
user    0m1.065s
sys     0m0.001s
$ time ./perf

real    0m1.063s
user    0m1.062s
sys     0m0.001s
$ time ./perf

real    0m1.072s
user    0m1.070s
sys     0m0.001s
```

Now update chk_job_request():
```
int
chk_job_request(int jobid, int *jobtype)
{
    int t;
    t = *jobtype = is_job_array(jobid);
    if (t != 0)
        t += jobid % 20;
    else
        t *= -1;
    if (t > 0)
        t *= 4;
    else
        t *= -5;
    return t;
}
```

And compile/run it again...
```
$ time ./perf

real    0m1.105s
user    0m1.104s
sys     0m0.000s
$ time ./perf

real    0m1.097s
user    0m1.095s
sys     0m0.001s
$ time ./perf

real    0m1.101s
user    0m1.099s
sys     0m0.001s
```

I like the suggestion from @suresh-thelkar

Reply:
@mike0042  I appreciate the effort taken to try validate my claim.

I hate to defend my claim below, but just for the sake of defending it.

I feel the optimization is way too feeble to show up in your tests.
Instead please have a look a below assembly level analysis

consider below file test.c (borrowed from your test code)
```
int
is_job_array(int jobid)
{
    return (jobid % 2);
}

int
chk_job_request_P(int jobid, int *jobtype)
{
    *jobtype = is_job_array(jobid);
    if (*jobtype != 0)
        *jobtype += jobid % 20;
    else
        *jobtype *= -1;
    if (*jobtype > 0)
        *jobtype *= 4;
    else
        *jobtype *= -5;
    return *jobtype;
}

int
chk_job_request_T(int jobid, int *jobtype)
{
    int t;
    t = *jobtype = is_job_array(jobid);
    if (t != 0)
        t += jobid % 20;
    else
        t *= -1;
    if (t > 0)
        t *= 4;
    else
        t *= -5;
    return t;
}
```
in the below output the second column shows the binary size of the functions considered
```
[shrini@pp479 opt_check]$ vi test.c 
[shrini@pp479 opt_check]$ gcc -c test.c 
[shrini@pp479 opt_check]$ nm test.o --radix=d --print-size --size-sort 
0000000000000000 0000000000000023 T is_job_array
0000000000000192 0000000000000123 T chk_job_request_T
0000000000000023 0000000000000169 T chk_job_request_P
[shrini@pp479 opt_check]$ gcc -c -O2 test.c 
[shrini@pp479 opt_check]$ nm test.o --radix=d --print-size --size-sort 
0000000000000000 0000000000000014 T is_job_array
0000000000000096 0000000000000068 T chk_job_request_T
0000000000000016 0000000000000076 T chk_job_request_P
[shrini@pp479 opt_check]$ 
```

We can see -O2 does a great job in reducing the size but still even with -O2 
 **chk_job_request_P** is 8 bytes of more text than **chk_job_request_T**

I have pasted below the comparison of disassembled text below if you are interested
![image](https://user-images.githubusercontent.com/30307577/37694930-32cf852a-2cf1-11e8-8e47-943427241c00.png)

Reply:
I think Mike and Suresh were pointing out the fact that you should prefer readability over negligible improvement in performance. But if you still wanna keep the local variable, then please make the following changes:
- Set the return pointer at the end of the function, setting it at the beginning is more error prone in my mind, what if the value changes and you(or another dev) only change the local variable and not update the return pointer.
- rename 't' to something more meaningful, this a general advice not related to any particular context.

Reply:
> I think Mike and Suresh were pointing out the fact that you should prefer readability over negligible improvement in performance.

I dont see such fact being pointed out in the comments. If it was so I would have considered removing 't'

>Set the return pointer at the end of the function, setting it at the beginning is more error prone in my mind, what if the value changes and you(or another dev) only change the local variable and not update the return pointer.

1. 't' is used as read only after its set and it wont change.
2. Lets say we set it at the end as you said, then what if you (or another dev) add a return in the middle and not update the return pointer.
3. Or what if somebody forgets to prefix '*' to 'jobtype'
4. Or what if somebody adds a call to nuclear armageddon.
5. Or what if  ... pls excuse me my imagination can go on 🤓 

>rename 't' to something more meaningful, this a general advice not related to any particular context.

1. 't' not introduced here, it already existed. please see LHS code.
2. 't' aptly stands for 'type'

Reply:
The code works fine either way, so I don't want to dwell on this more than need be. I will say this... a smaller function doesn't always mean a faster function. The fact that the P version performs better than the T version is evidence of this, but the difference is negligible. The compiler is structuring the assembly to take advantage of features such as pipelining.

My preference would be to use the P version because it is more readable, IMHO, but I'm not going to insist upon it.

Reply:
Addressed in rc 19 https://github.com/PBSPro/pbspro/pull/590/commits/5e926b0511c16ea932ca1ed7570e997d849255eb